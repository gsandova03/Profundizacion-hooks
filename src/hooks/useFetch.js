import { useEffect, useRef, useState } from 'react'

export const useFetch = ( url ) => {

  const isMounted = useRef(true);
  const [state, setState] = useState({
    data: null,
    loading: true,
    error: null
  });

  useEffect(() => {
    return () => {
      //Cambiar el valor de nuestra referencia, cuando el componente se desmonte
      isMounted.current = false;
    }
  }, [])

  // const getData = async() => {
  //   await fetch( url )
  //     .then( res => res.json() )
  //     .then( datos => setState({
  //       data: datos,
  //       loading: false,
  //       error: null
  //     }) )
  // }

  // useEffect(() => {
  //   getData();
  // }, [ url ])

  useEffect(() => {

    setState({
      data: null,
      loading: true,
      error: null
    });

    fetch( url )
      .then( res => res.json() )
      .then( data => {

        //condicionar el setState para evitar la fuga de momoria o la asignacion del esatado, cuando el componente este desmontado
        if( isMounted.current ){
          setState({
            data: data,
            loading: false,
            error: null
          })
        }else{
          console.log('setState no se llamo')
        }
      })
  }, [ url ])

  return state;
}
