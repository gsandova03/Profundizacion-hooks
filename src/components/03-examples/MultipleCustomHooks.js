import React from 'react'
import { useCounter } from '../../hooks/useCounter';
import { useFetch } from '../../hooks/useFetch'

export const MultipleCustomHooks = () => {

  const { counter, increment, decrement, reset } = useCounter(1);

  const { loading, data } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);
  const { author, quote } = !loading && data[0];

  return (
    <div>
      <h1>Customs Hooks</h1>
      <hr/>
        { loading 
          ? 
            (
              <div className=' alert alert-info text-center'>
                Loading ...
              </div> 
            )
          :
            (
              <blockquote className="m-5 blockquote text-right">
                <p><i className="quote">{ quote }</i></p>
                <p><i className="author">{ author }</i></p>
              </blockquote>
            )
        }
        {/* {
          counter <= 1
          ?(
            <button disabled className="btn btn-primary m-1">Ant quote</button>      
          )

          :(
            <button onClick={ decrement } className="btn btn-primary m-1">Ant quote</button>
          )
        } */}
      {/* <button onClick={ reset } className="btn btn-primary m-1">Reset</button> */}
      <button onClick={ increment } className="btn btn-primary m-1">Sig quote</button>
    </div>
  )
}
