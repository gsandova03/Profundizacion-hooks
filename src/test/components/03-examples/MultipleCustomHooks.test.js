import React from 'react';
import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';

//implementacion de mock, lo cual permite evitar llamados a nuestra funcion y nos permite agregar valores de pruebas
jest.mock( '../../../hooks/useFetch' );
// jest.mock( '../../../hooks/useCounter' );

describe('Pruebas en el componente <MultipleCustomHooks/>', () => {
  
  // useCounter.mockReturnValue({
  //   counter: 1,
  //   increment: () => {} 
  // })

  test('debe de renderizarce correctamente', () => {
    
    //mock para fingir que nuestro componente tiene informacion
    //Agregar valores simulados
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null
    })

    const wrapper = shallow( <MultipleCustomHooks /> )

    expect( wrapper ).toMatchSnapshot();

  })

  test('debe de mostrar la informacion', () => {
    useFetch.mockReturnValue({
      data: [{
        author: "gabriel",
        quote: "Hola mundo"
      }],
      loading: false,
      error: null
    })

    const wrapper = shallow( <MultipleCustomHooks /> )

    expect( wrapper.find('.alert').exists() ).toBe( false )
    expect( wrapper.find('.quote').text().trim() ).toBe( "Hola mundo" )
    expect( wrapper.find('.author').text().trim() ).toBe( "gabriel" )
    
  })
  
})
