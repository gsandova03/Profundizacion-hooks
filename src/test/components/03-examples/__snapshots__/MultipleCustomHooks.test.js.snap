// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <MultipleCustomHooks/> debe de renderizarce correctamente 1`] = `
<div>
  <h1>
    Customs Hooks
  </h1>
  <hr />
  <div
    className=" alert alert-info text-center"
  >
    Loading ...
  </div>
  <button
    className="btn btn-primary m-1"
    onClick={[Function]}
  >
    Sig quote
  </button>
</div>
`;
